src/util/query_exec_util.cpp

```diff
@@ -0,0 +1,283 @@
+  }
+
+  auto accessor = catalog_->GetAccessor(txn, db_oid_, DISABLED);
+  auto result = PlanStatement(query, nullptr, nullptr);
+  const std::unique_ptr<network::Statement> &statement = result.first;
+  const std::unique_ptr<planner::AbstractPlanNode> &out_plan = result.second;
+  NOISEPAGE_ASSERT(!network::NetworkUtil::DMLQueryType(statement->GetQueryType()), "ExecuteDDL expects DDL statement");
+
+  // Handle SET queries
+  bool status = true;
+  if (statement->GetQueryType() == network::QueryType::QUERY_SET) {
+    const auto &set_stmt = statement->RootStatement().CastManagedPointerTo<parser::VariableSetStatement>();
+    settings_->SetParameter(set_stmt->GetParameterName(), set_stmt->GetValues());
+    status = true;
+  } else {
+    switch (statement->GetQueryType()) {
+      case network::QueryType::QUERY_CREATE_TABLE:
+        status = execution::sql::DDLExecutors::CreateTableExecutor(
+            common::ManagedPointer<planner::CreateTablePlanNode>(
+                reinterpret_cast<planner::CreateTablePlanNode *>(out_plan.get())),
+            common::ManagedPointer(accessor), db_oid_);
+        break;
+      case network::QueryType::QUERY_DROP_INDEX:
+        // Drop index does not need execution of compiled query
+        status =
+            execution::sql::DDLExecutors::DropIndexExecutor(common::ManagedPointer<planner::AbstractPlanNode>(out_plan)
+                                                                .CastManagedPointerTo<planner::DropIndexPlanNode>(),
+                                                            common::ManagedPointer<catalog::CatalogAccessor>(accessor));
+        break;
+      case network::QueryType::QUERY_CREATE_INDEX:
+        // TODO(lin): We actually don't need to populate the index tuples after creating the index placeholder for the
+        //  "what-if" API. But since we need to execute the query to get the features, we need to compile and execute
+        //  the query for now.
+        status = execution::sql::DDLExecutors::CreateIndexExecutor(
+            common::ManagedPointer<planner::AbstractPlanNode>(out_plan)
+                .CastManagedPointerTo<planner::CreateIndexPlanNode>(),
+            common::ManagedPointer<catalog::CatalogAccessor>(accessor));
+        break;
+      default:
+        NOISEPAGE_ASSERT(false, "Unsupported QueryExecUtil::ExecuteStatement");
+        break;
+    }
+  }
+
+  if (require_commit) {
+    // Commit if success
+    EndTransaction(status);
+  }
+
+  return status;
+}
+
+size_t QueryExecUtil::CompileQuery(const std::string &statement,
+                                   common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+                                   common::ManagedPointer<std::vector<type::TypeId>> param_types, bool *success) {
+  auto txn = common::ManagedPointer<transaction::TransactionContext>(txn_);
+  bool require_commit = false;
+  if (txn == nullptr) {
```

Comment:
This code appears a couple of times, maybe it can be refactored into an internal helper function